libnavigation-core/src/test/java/com/mapbox/navigation/core/routerefresh/PlannedRouteRefreshIntegrationTest.kt

```diff
@@ -0,0 +1,296 @@
+            successfulAttemptNumber = 0
+        )
+
+        routeRefreshController.onRoutesChanged(
+            RoutesUpdatedResult(newRoutes, emptyList(), RoutesExtra.ROUTES_UPDATE_REASON_NEW)
+        )
+
+        testDispatcher.advanceTimeBy(40_000)
+        assertEquals(0, refreshObserver.refreshes.size)
+
+        testDispatcher.advanceTimeBy(20_000)
+
+        assertEquals(1, refreshObserver.refreshes.size)
+        assertEquals(
+            newRoutes[0].id,
+            refreshObserver.refreshes.first().primaryRouteRefresherResult.route.id
+        )
+        assertEquals(
+            newRoutes[1].id,
+            refreshObserver.refreshes.first().alternativesRouteRefresherResults.first().route.id
+        )
+        assertEquals(
+            listOf(
+                RouteRefreshExtra.REFRESH_STATE_STARTED,
+                RouteRefreshExtra.REFRESH_STATE_CANCELED,
+                RouteRefreshExtra.REFRESH_STATE_STARTED,
+                RouteRefreshExtra.REFRESH_STATE_FINISHED_SUCCESS,
+            ),
+            stateObserver.getStatesSnapshot()
+        )
+    }
+
+    @Test
+    fun routeRefreshWhenRoutesAreCleared() = runBlocking {
+        val routes = setUpRoutes("route_response_route_refresh_multileg.json")
+
+        routeRefreshController = createRefreshController(30_000)
+        routeRefreshController.registerRouteRefreshObserver(refreshObserver)
+        routeRefreshController.registerRouteRefreshStateObserver(stateObserver)
+
+        routeRefreshController.onRoutesChanged(
+            RoutesUpdatedResult(routes, emptyList(), RoutesExtra.ROUTES_UPDATE_REASON_NEW)
+        )
+
+        testDispatcher.advanceTimeBy(20_000)
+
+        routeRefreshController.onRoutesChanged(
+            RoutesUpdatedResult(emptyList(), emptyList(), RoutesExtra.ROUTES_UPDATE_REASON_CLEAN_UP)
+        )
+
+        testDispatcher.advanceTimeBy(30_000)
+        assertEquals(0, refreshObserver.refreshes.size)
+        assertEquals(0, stateObserver.getStatesSnapshot().size)
+    }
+
+    @Test
+    fun routesChangeToAlternativeWhenRequestIsInProgress() = runBlocking {
+        val routes = setUpRoutes("route_response_route_refresh_multileg.json", responseDelay = 3000)
+        val initialRoutes = routes.take(1)
+        routeRefreshController = createRefreshController(40_000)
+        routeRefreshController.registerRouteRefreshObserver(refreshObserver)
+        routeRefreshController.registerRouteRefreshStateObserver(stateObserver)
+
+        routeRefreshController.onRoutesChanged(
+            RoutesUpdatedResult(initialRoutes, emptyList(), RoutesExtra.ROUTES_UPDATE_REASON_NEW)
+        )
+
+        testDispatcher.advanceTimeBy(42_000) // start the refresh request
+
+        // set alternatives
+        routeRefreshController.onRoutesChanged(
+            RoutesUpdatedResult(routes, emptyList(), RoutesExtra.ROUTES_UPDATE_REASON_ALTERNATIVE)
+        )
+
+        testDispatcher.advanceTimeBy(6_000) // response delay x routes.size
```

Comment:
```suggestion
        testDispatcher.advanceTimeBy(3_000) // response delay x routes.size
```
or even
```suggestion
        testDispatcher.advanceTimeBy(responceDelay)
```
3 seconds is enough there, because unlike instrumentation tests, where mock http server process requests one by one and delay stops it, fake router processes requests in parallel